### PR TITLE
[ZEPPELIN-2217] AdvancedTransformation for Visualization

### PR DESCRIPTION
**Pull request by 1ambda**
### What is this PR for?

`AdvancedTransformation` has more detailed options while providing existing features of `PivotTransformation` and `ColumnselectorTransformation` which Zeppelin already has 

av_in_30sec

Here are some features which advanced-transformation can provide.

1. **(screenshot)** multiple sub charts
2. **(screenshot)** parameter widgets: `input`, `checkbox`, `option`, `textarea`
3. **(screenshot)** expand/fold axis and parameter panels
4. **(screenshot)** clear axis and parameter panels
5. **(screenshot)** remove duplicated columns in an axis
6. **(screenshot)** limit column count in an axis
7. configurable char axes: `valueType`, `axisType`, `description`, ...
8. configurable chart parameters
9. lazy transformation
10. parsing parameters automatically based on their type: `int`, `float`, `string`, `JSON`
11. multiple transformation methods
12. re-initialize whole configuration based on spec hash.
13. **(screenshot)** shared axis

#### API Details: Spec

`AdvancedTransformation` requires `spec` which includes axis and parameter details for charts.

- Let's create 2 sub-charts called `simple-line` and `step-line`.
- Each sub chart can have different `axis` and `parameter` depending on their requirements.

```js
  constructor(targetEl, config) {
    super(targetEl, config)

    const spec = {
      charts: {
        'simple-line': {
          sharedAxis: true, /** set if you want to share axes between sub charts, default is `false` */
          axis: {
            'xAxis': { dimension: 'multiple', axisType: 'key', },
            'yAxis': { dimension: 'multiple', axisType: 'aggregator'},
            'category': { dimension: 'multiple', axisType: 'group', },
          },
          parameter: {
            'xAxisUnit': { valueType: 'string', defaultValue: '', description: 'unit of xAxis', },
            'yAxisUnit': { valueType: 'string', defaultValue: '', description: 'unit of yAxis', },
            'dashLength': { valueType: 'int', defaultValue: 0, description: 'the length of dash', },
          },
        },

        'step-line': {
          axis: {
            'xAxis': { dimension: 'single', axisType: 'unique', },
            'yAxis': { dimension: 'multiple', axisType: 'value', },
          },
          parameter: {
            'xAxisUnit': { valueType: 'string', defaultValue: '', description: 'unit of xAxis', },
            'yAxisUnit': { valueType: 'string', defaultValue: '', description: 'unit of yAxis', },
            'noStepRisers': { valueType: 'boolean', defaultValue: false, description: 'no risers in step line', widget: 'checkbox', },
        },

      },
    }

    this.transformation = new AdvancedTransformation(config, spec)
  }
```

####  API Details: Axis Spec

| Field Name | Available Values (type) | Description |
| --- | --- | --- |
|`dimension` | `single` | Axis can contains only 1 column |
|`dimension` | `multiple` | Axis can contains multiple columns |
|`axisType` | `key` | Column(s) in this axis will be used as `key` like in `PivotTransformation`. These columns will be served in `column.key` |
|`axisType` | `aggregator` | Column(s) in this axis will be used as `value` like in `PivotTransformation`. These columns will be served in `column.aggregator` |
|`axisType` | `group` | Column(s) in this axis will be used as `group` like in `PivotTransformation`. These columns will be served in `column.group` |
|`axisType` | (string) | Any string value can be used here. These columns will be served in `column.custom` |
|`maxAxisCount` | (int) | The maximum column count that this axis can contains. (unlimited if `undefined`) |
|`valueType` | (string) | Describe the value type just for annotation |

Here is an example.

```js
          axis: {
            'xAxis': { dimension: 'multiple', axisType: 'key',  },
            'yAxis': { dimension: 'multiple', axisType: 'aggregator'},
            'category': { dimension: 'multiple', axisType: 'group', maxAxisCount: 2, valueType: 'string', },
          },
```

####  API Details: Parameter Spec

| Field Name | Available Values (type) | Description |
| --- | --- | --- |
|`valueType` | `string` | Parameter which has string value |
|`valueType` | `int` | Parameter which has int value |
|`valueType` | `float` | Parameter which has float value |
|`valueType` | `boolean` | Parameter which has boolean value used with `checkbox` widget usually |
|`valueType` | `JSON` | Parameter which has JSON value used with `textarea` widget usually. `defaultValue` should be `""` (empty string). This ||`defaultValue` | (any) | Default value of this parameter. `JSON` type should have `""` (empty string) |
|`description` | (string) | Description of this parameter. This value will be parsed as HTML for pretty output |
|`widget` | `input` |  Use [input](https://developer.mozilla.org/en/docs/Web/HTML/Element/input) widget. This is the default widget (if `widget` is undefined)|
|`widget` | `checkbox` |  Use [checkbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox) widget. |
|`widget` | `textarea` |  Use [textarea](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea) widget. |
|`widget` | `option` |  Use [select + option](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select) widget. This parameter should have `optionValues` field as well. |
|`optionValues` | (Array<string>) |  Available option values used with the `option` widget |

Here is an example.

```js
parameter: {
  // string type, input widget
  'xAxisUnit': { valueType: 'string', defaultValue: '', description: 'unit of xAxis', },

  // boolean type, checkbox widget
  'inverted': { widget: 'checkbox', valueType: 'boolean', defaultValue: false, description: 'invert x and y axes', },

  // string type, option widget with `optionValues`
  'graphType': { widget: 'option', valueType: 'string', defaultValue: 'line', description: 'graph type', optionValues: [ 'line', 'smoothedLine', 'step', ], },

  // HTML in `description`
  'dateFormat': { valueType: 'string', defaultValue: '', description: 'format of date (<a href="https://docs.amcharts.com/3/javascriptcharts/AmGraph#dateFormat">doc</a>) (e.g YYYY-MM-DD)', },
 
  // JSON type, textarea widget
  'yAxisGuides': { widget: 'textarea', valueType: 'JSON', defaultValue: '', description: 'guides of yAxis ', },
```

#### API Details: Transformer Spec

`AdvancedTransformation` supports 3 transformation methods. The return value will depend on the transformation method type. 

```js
    const spec = {
      charts: {
        'simple': {
          /** default value of `transform.method` is the flatten cube.  */
          axis: { ... },
          parameter: { ... }
        },

        'cube-group': {
          transform: { method: 'cube', },
          axis: { ... },
          parameter: { ... },
        }

        'no-group': {
          transform: { method: 'raw', },
          axis: { ... },
          parameter: { ... },
        }
```

| Field Name | Available Values (type) | Description |
| --- | --- | --- |
|`method` | `object` |  designed for [amcharts: serial](https://www.amcharts.com/demos/date-based-data/) | 
|`method` | `array` | designed for [highcharts: column](http://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/demo/column-basic/) |
|`method` | `drill-down` | designed for [highcharts: drill-down](http://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/demo/column-drilldown/) |
|`method` | `raw` | will return the original `tableData.rows` | 

Whatever you specified as `transform.method`, the `transformer` value will be always function for lazy computation. 

```js
// advanced-transformation.util#getTransformer

  if (transformSpec.method === 'raw') {
    transformer = () => { return rows; }
  } else if (transformSpec.method === 'array') {
    transformer = () => {
      ...
      return { ... }
    }
  }

```

#### Feature Details: Automatic parameter parsing

Advanced transformation will parse parameter values automatically based on their type: `int`, `float`, `string`, `JSON`

- See also `advanced-transformation-util.js#parseParameter`

#### Feature Details: re-initialize the whole configuration based on spec hash

```js
// advanced-transformation-util#initializeConfig

  const currentVersion = JSON.stringify(spec)
  if (!config.spec || !config.spec.version || config.spec.version !== currentVersion) {
    spec.version = currentVersion
    // reset config...
  }
```

#### Feature Details: Shared Axes

If you set `sharedAxis` to `true` in chart specification, then these charts will share their axes. (default is `false`)

```js
    const spec = {
      charts: {
        'column': {
          transform: { method: 'array', },
          sharedAxis: true,
          axis: { ... },
          parameter: { ... },
        },

        'stacked': {
          transform: { method: 'array', },
          sharedAxis: true,
          axis: { ... }
          parameter: { ... },
        },
```

sharedaxis

#### API Details: Usage in Visualization#render()

Let's assume that we want to create 2 sub-charts called `basic` and `no-group`.

- https://github.com/1ambda/zeppelin-ultimate-line-chart (an practical example)

```js
  drawBasicChart(parameter, column, transformer) { 
    const { ... } = transformer()
  }

  drawNoGroupChart(parameter, column, transformer) { 
    const { ... } = transformer()
  }

  render(data) {
    const { chart, parameter, column, transformer, } = data

    if (chart === 'basic') {
      this.drawBasicChart(parameter, column, transformer)
    } else if (chart === 'no-group') {
      this.drawNoGroupChart(parameter, column, transformer)
    }
  }
```

### What type of PR is it?
[Feature]

### Todos

NONE

### What is the Jira issue?

[ZEPPELIN-2217](https://issues.apache.org/jira/browse/ZEPPELIN-2217)

### How should this be tested?

1. Clone https://github.com/1ambda/zeppelin-ultimate-line-chart
2. Create a symbolic link `ultimate-line-chart.json` into `$ZEPPELIN_HOME/helium`
3. Modify the `artifact` value to proper absolute path considering your local machine.
4. Install the above visualization in `localhost:9000/#helium`
5. Test it

#### 1. *(screenshot)* multiple sub charts

av_multiple_charts

#### 2. *(screenshot)* parameter widgets: `input`, `checkbox`, `option`, `textarea`

av_widgets_new

#### 3. *(screenshot)* expand/fold axis and parameter panels

av_fold_expand

#### 4. *(screenshot)* clear axis and parameter panels

av_clean_buttons

#### 5. *(screenshot)* remove duplicated columns in an axis

av_duplicated_columns

#### 6. *(screenshot)* limit column count in an axis

av_maxaxiscount

### Questions:
* Does the licenses files need update? - NO
* Is there breaking changes for older versions? - NO
* Does this needs documentation? - NO

### CODE REVIEW

**Comment by AhyoungRyu** [comment]
@1ambda Cool! Let me test this and I will get back you soon.

**Comment by 1ambda** [comment]
@AhyoungRyu Thanks! 

I didn't attach screenshots for examples which use `aggregator: false,`.  I will prepare in few days.

**Comment by 1ambda** [comment]
Some features, improvements which can be done in this PR

- refine `transform()` return value
- selectable dynamic parameter. `{ ..., name: "theme" paramType: "optional", options: ["light", "dark"]`
- support multiple chart types in one visualization. `{ subcharts: ["area", "stacked", "stream"] }`
- customized axis spec + dynamic parameters for each sub chart type.
- parse dynamic parameters based on types (`string`, `number`, `JSON`)

**Comment by AhyoungRyu** [comment]
@1ambda Okay then feel free to ping me when they're done :)

**Comment by 1ambda** [comment]
CI passed. I will update PR description

**Comment by 1ambda** [comment]
I'v just updated the PR description. Please help review this @AhyoungRyu :)

**Comment by 1ambda** [comment]
Some ideas

- multiple aggregaor axes support
- remove aggregator column name from the legend when only 1 aggr column and group is enabled

**Comment by Leemoonsoo** [comment]
@1ambda Great work!

Let me test this interesting feature in this weekend.

**Comment by 1ambda** [comment]
updated

Please check https://github.com/1ambda/zeppelin-ultimate-line-chart as well.

Thanks!

**Comment by 1ambda** [comment]
Few things are left. Thus i'v just changed to WIP again. Sorry for confusing.

- adding more `transformation.method` for columnar and drill-down charts: `array:key`, `array:group`, `array:drilldown`
- fixing bugs.

The final output will be served with 2 examples

1. column charts including drill-down functionality
2. pie charts including drill-down functionality

Let me mention again after then :)

**Comment by 1ambda** [comment]
@Leemoonsoo, @AhyoungRyu implementation is finally finished (IMO 😭 )

- Please help review this. 
- Testing these 2 examples will be helpful (not published yet) [ultimate-line-chart](https://github.com/1ambda/zeppelin-ultimate-line-chart), [ultimate-column-chart](https://github.com/1ambda/zeppelin-ultimate-column-chart)
- I will write documentation based on the PR description if review is done. I would like to postpone writing doc because it can change easily.

Here is GIF for drill-down and shared axis what i added in last few commits :)

drilldown

**Comment by AhyoungRyu** [comment]
@1ambda Looks awesome! Let me test it and i'll get back to you soon.

**Comment by Leemoonsoo** [comment]
@1ambda Great work!!!

Looks great to me!

I think it's worth to move usages and examples in this PR description under `/docs`. 
Where do you think this information can be placed?

We didn't have any unittest for pivot transformation. So it was hard to improve / change without any side effects. So having some basic unittest for advanced transformation will help other people add and improve the feature much faster.

**Comment by 1ambda** [comment]
Let me add docs and unit tests. I guess it will take few days. Then i will comment. Thanks @Leemoonsoo

**Comment by AhyoungRyu** [comment]
@1ambda I feel so sorry for my late response. Tested this patch \w your Helium package: zeppelin-ultimate-line-chart. Here are some thought and question(?) in my head. 
 
1. "Save" & "Delete" button on top of the setting table
 - It'll be better there are some feedback after clicking these buttons like "Changes are successfully saved" or "Changes will be discarded". How about small `ngtoast` or bootstrap dialog? 
small_btns
Probably ppl can click the trash can button by mistake. Then they'll lose the setting. To prevent this, we can provide some alert for this. 

 - I think we can hide "Save" & "Delete" button when the setting table is folded. A good example is in our interpreter setting page. As you know, users only can save/remove the setting when they click "edit" button. So how about activating those disket & trash btns only when the setting table is expanded? 

2. no-group field 
 - What is difference between `no-group` and just letting `group` field as blank? Is there special reason for creating `no-group` field? (It's question actually :D)

 - `xAxis` and `yAxis` field size are different in `no-group` section.
screen shot 2017-03-29 at 2 25 50 pm

And again, this is really cool!

**Comment by 1ambda** [comment]
@AhyoungRyu 

Thanks for detailed review. 

1. It would be nice to have some popup or feedback when conf is updated.
2. Let me hide them when panel is minimized (e.g the chart panel, the parameter panel)

Regarding to `3. no-group field`, it's quite hard to say in just few lines. So let me describe more.

(A) Some charts require aggregation. (= groupBy in SQL)
(B) But some charts might not need aggregation. Assume that it's already grouped by SQL

In ultimate-line-chart, there are sub charts called `line`, `dashed`, `step` and `no-group`.

- `line`, `dashed`,` step` uses aggregation like zeppelin's existing charts. (similar to `PivotTransformation`)
- But `no-group` doesn't use aggregation. (similar to `ColumnTransformation`)

**Comment by AhyoungRyu** [comment]
@1ambda Appreciate for such detailed explanation! 
Regarding `no-group`, I can assume the other ppl can be confused like me. So it'll be better to add some description for that to the documentation. (you already wrote all needed contents in your comments :D)

FYI I suggested to separate Helium related docs page from "Contribute" section in the dropdown menu and create new section "Helium Framework" in #2202(Helium is worth to have its own kk). So your new docs page "Writing Zeppelin Visualization: Transformation(or another name?)" can be put under this section.
And maybe more self-descriptive name would be better instead of just `Transformation`(hard to recognize at a glance what kind of transformation is). Something like `Writing Zeppelin Visualization: Advanced Chart Configuration`.. 
But I don't wanna hang this PR because of docs. Surely you can handle it in another PR.

**Comment by 1ambda** [comment]
@AhyoungRyu 

1. Hide other buttons when panel is minified

image

2. Added doc to `docs/development/writingzeppelinvisualization_transformation.md`. It looks like 

2098_docs

3. Regarding to popup, i think we can skip it because if parameter is applied or reset, the chart will change according to the options. The chart will remain same only if parameter is not changed at all and user can recognize when he/she does't change parameter.

@Leemoonsoo 

4. I added many tests to `advanced-transformation-util.test.js`. It covers most cases and many functions. Here is the file.

- https://github.com/1ambda/zeppelin/blob/6a0559b18542c60b1211095f7583a71e5fb0c136/zeppelin-web/src/app/tabledata/advanced-transformation-util.test.js

**Comment by 1ambda** [comment]
CI status: https://travis-ci.org/1ambda/zeppelin/builds/218186828

**Comment by 1ambda** [comment]
@AhyoungRyu Updated docs as well.

image

**Comment by AhyoungRyu** [comment]
@1ambda Thanks for letting me know. Just look through what you said, generally looks good. Let me test again.

File: docs/development/writingzeppelinvisualization_transformation.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+## Overview 
+
+Transformations 
+
+- **renders** setting which allows users to set columns and 
+- **transforms** table rows according to the configured columns.
+
+Zeppelin provides 4 types of transformations.
+
+## 1. PassthroughTransformation
```
Comment (1ambda) [inline docs/development/writingzeppelinvisualization_transformation.md]: sure.

File: docs/development/writingzeppelinvisualization_transformation.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+See [pivot.js](https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/app/tabledata/pivot.js)
+
+## 4. AdvancedTransformation
+
+`AdvancedTransformation` has more detailed options while providing existing features of `PivotTransformation` and `ColumnselectorTransformation`
+
+- multiple sub charts
+- configurable chart axes
+- parameter widgets: `input`, checkbox`, `option`, `textarea`
```
Comment (1ambda) [inline docs/development/writingzeppelinvisualization_transformation.md]: will be fixed.

File: docs/development/writingzeppelinvisualization_transformation.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+axis: {
+  'xAxis': { dimension: 'multiple', axisType: 'key',  },
+  'yAxis': { dimension: 'multiple', axisType: 'aggregator'},
+  'category': { dimension: 'multiple', axisType: 'group', maxAxisCount: 2, valueType: 'string', },
+},
+```
+
+<br/>
+
+#### Spec: `sharedAxis`
```
Comment (1ambda) [inline docs/development/writingzeppelinvisualization_transformation.md]: will be fixed.

File: docs/development/writingzeppelinvisualization_transformation.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+Here is actual usage.
+
+```js
+class AwesomeVisualization extends Visualization {
+  constructor(...) { /** setup your spec */ }
+  
+  ... 
+  
+  // `render` will be called whenever `axis` or `parameter` are changed
```
Comment (1ambda) [inline docs/development/writingzeppelinvisualization_transformation.md]: will be fixed.

File: docs/development/writingzeppelinvisualization_transformation.md (hunk 1 above)
Comment (1ambda) [inline docs/development/writingzeppelinvisualization_transformation.md]: @AhyoungRyu I added descriptions for existing transformations. But quite hard to explain `ColumnselectorTransformation`. So please tell me if there are more appropriate words for that.

**Comment by 1ambda** [comment]
@AhyoungRyu resolved all comments. Thanks for detailed review.

**Comment by AhyoungRyu** [comment]
@1ambda Looks great! 
So sorry I forgot to mention that there is another place to be updated regarding [this change](https://github.com/apache/zeppelin/pull/2098#issuecomment-291390619). 



Could you update [`docs/index.html`](https://github.com/apache/zeppelin/blob/master/docs/index.md#more) as well?

**Comment by 1ambda** [comment]
image

@AhyoungRyu I forgot as well to modify `index.md` 😭.
I'v just updated.

**Comment by AhyoungRyu** [comment]
Cool! LGTM. Looking forward to many usage of this advanced transformation by Zeppelin-Helium fans :)

**Comment by 1ambda** [comment]
Found few bugs. Let me update more.

**Comment by 1ambda** [comment]
Few things are fixed. And really ready to be reviewed.

- Use `description` field in `axisType` as description will be displayed using tooltip in 7c6768f
- Fix yarn cache issue in c91a033, 4887800
- Add `array:2-key` transform method which is designed for xyz chart (e.g bubble) in bf56761

@AhyoungRyu @Leemoonsoo Please help review this. Here is a pretty chart you might be interested in using newly added transform method :)

scatter
bubble_1
bubble_2

**Comment by 1ambda** [comment]
Removed `WIP` from PR title.

**Comment by Leemoonsoo** [comment]
Tremendous work @1ambda !
LGTM and merge to master if no further discussions.

**Pull request closed without merging**